libr/bin/dwarf.c

```diff
@@ -326,15 +326,65 @@ static const char *dwarf_unit_types[] = {
 	[DW_UT_hi_user] = "DW_UT_hi_user",
 };
 
+enum {
+	DWARF_SN_ABBREV,
+	DWARF_SN_INFO,
+	DWARF_SN_FRAME,
+	DWARF_SN_LINE,
+	DWARF_SN_LOC,
+	DWARF_SN_STR,
+	DWARF_SN_LINE_STR,
+	DWARF_SN_RANGES,
+	DWARF_SN_ARANGES,
+	DWARF_SN_PUBNAMES,
+	DWARF_SN_PUBTYPES,
+
+	DWARF_SN_MAX
+};
+
+static const char *dwarf_sn_elf[DWARF_SN_MAX] = {
+	[DWARF_SN_ABBREV] = "debug_abbrev",
+	[DWARF_SN_INFO] = "debug_info",
+	[DWARF_SN_FRAME] = "debug_frame",
+	[DWARF_SN_LINE] = "debug_line",
+	[DWARF_SN_LOC] = "debug_loc",
+	[DWARF_SN_STR] = "debug_str",
+	[DWARF_SN_LINE_STR] = "debug_line_str",
+	[DWARF_SN_RANGES] = "debug_ranges",
+	[DWARF_SN_ARANGES] = "debug_aranges",
+	[DWARF_SN_PUBNAMES] = "debug_pubnames",
+	[DWARF_SN_PUBTYPES] = "debug_pubtypes",
+};
+
+/* XXX: xcoff64 discovers DWARF sections by SSUBTYP_DW{...}, not by name */
+static const char *dwarf_sn_xcoff64[DWARF_SN_MAX] = {
+	[DWARF_SN_ABBREV] = "dwabrev",
+	[DWARF_SN_INFO] = "dwinfo",
+	[DWARF_SN_FRAME] = "dwframe",
+	[DWARF_SN_LINE] = "dwline",
+	[DWARF_SN_LOC] = "dwloc",
+	[DWARF_SN_RANGES] = "dwrnges",
+	[DWARF_SN_ARANGES] = "dwarnge",
+	[DWARF_SN_STR] = "dwstr", /* XXX: unverified */
+	[DWARF_SN_PUBNAMES] = "dwpbnms",
+	[DWARF_SN_PUBTYPES] = "dwpbtyp"
+};
+
 // 1 of 20s spent in this non-mnemonized function
-static RBinSection *getsection(RBin *a, const char *sn) {
+static RBinSection *getsection(RBin *a, int sn) {
 	RListIter *iter;
 	RBinSection *section = NULL;
 	RBinFile *binfile = a ? a->cur: NULL;
 	RBinObject *o = binfile ? binfile->bo : NULL;
 	if (R_LIKELY (o && o->sections)) {
+		/* XXX: xcoff64 specific hack */
+		const char * const *name_tab = strcmp (o->info->rclass, "xcoff64")
+			? dwarf_sn_elf
+			: dwarf_sn_xcoff64;
+		assert (sn >= 0 && sn < DWARF_SN_MAX);
+		const char *name_str = name_tab[sn];
 		r_list_foreach (o->sections, iter, section) {
-			if (strstr (section->name, sn)) {
+			if (strstr (section->name, name_str)) {
 				if (strstr (section->name, "zdebug")) {
```

Comment:
oh i think that should be a good thing to have at some point too. ive seen those compressed dwarf sections more freqntly these days, but not for this pr. anyway i dont see why your pr is breaking so much tests. can you rebase and see if the magic happens? because your changes look good to me

Reply:
> Anyway i dont see why your pr is breaking so much tests

Fixed 🙏🏻